cmd/juju/common/cloudcredential.go

```diff
@@ -50,11 +50,17 @@ func GetOrDetectCredential(
 	// the environment.
 	ctx.Verbosef("no credentials found, checking environment")
 
-	detected, err := modelcmd.DetectCredential(args.Cloud.Name, provider)
-	if errors.Cause(err) == modelcmd.ErrMultipleCredentials {
+	detected, selected, detectErr := modelcmd.DetectCredential(args.Cloud.Name, args.CredentialName, provider)
+	// Attempt to detect a credential if the credential name is not empty, but
+	// detect credential doesn't return a selected credential based on the
+	// credential name.
+	if !selected && args.CredentialName != "" {
+		return credential, chosenCredentialName, regionName, false, err
+	}
+	if errors.Cause(detectErr) == modelcmd.ErrMultipleCredentials {
```

Comment:
This could be jsut
```
if detectErr != nil {
    if errors.Cause(detectErr) == modelcmd.ErrMultipleCredentials {
        return fail(ErrMultipleDetectedCredentials)
    }
    return fail(errors.Trace(detectErr))
}
```